pkg/controllers/management/clusterprovisioner/helm_cleaner.go

```diff
@@ -0,0 +1,230 @@
+
+		if mappingData.DeprecatedInVersion != "" {
+			apiVersion = mappingData.DeprecatedInVersion
+		} else {
+			apiVersion = mappingData.RemovedInVersion
+		}
+
+		if !semver.IsValid(apiVersion) {
+			return "", errors.Errorf("Failed to get the deprecated or removed Kubernetes version for API: %s", strings.ReplaceAll(deprecatedAPI, "\n", " "))
+		}
+
+		if count := strings.Count(manifest, deprecatedAPI); count > 0 {
+			if semver.Compare(apiVersion, kubeVersion) > 0 {
+				logrus.Printf("The following API does not require mapping as the "+
+					"API is not deprecated or removed in Kubernetes '%s':\n\"%s\"\n", apiVersion,
+					deprecatedAPI)
+			} else {
+				if supportedAPI == "" {
+					logrus.Printf("Found %d instances of deprecated or removed Kubernetes API:\n\"%s\"\n", count, deprecatedAPI)
+
+					for repl := 0; repl < count; repl++ {
+						// find the position where the API header is
+						apiIndex := strings.Index(manifest, deprecatedAPI)
+
+						// find the next separator index
+						separatorIndex := strings.Index(manifest[apiIndex:], "---\n")
+
+						// find the previous separator index
+						previousSeparatorIndex := strings.LastIndex(manifest[:apiIndex], "---\n")
+
+						/*
+						 * if no previous separator index was found, it means the resource is at the beginning and not
+						 * prefixed by ---
+						 */
+						if previousSeparatorIndex == -1 {
+							if apiIndex == 0 {
+								previousSeparatorIndex = 0
+							} else {
+								previousSeparatorIndex = apiIndex - 1
+							}
+						}
+
+						if separatorIndex == -1 { // this means we reached the end of input
+							manifest = manifest[:previousSeparatorIndex]
+						} else {
+							manifest = manifest[:previousSeparatorIndex] + manifest[separatorIndex+apiIndex:]
+						}
+					}
+
+					manifest = strings.Trim(manifest, "\n")
+				} else {
+					logrus.Printf("Found %d instances of deprecated or removed Kubernetes API:\n\"%s\"\nSupported API equivalent:\n\"%s\"\n", count, deprecatedAPI, supportedAPI)
+					manifest = strings.ReplaceAll(manifest, deprecatedAPI, supportedAPI)
+				}
+			}
+		}
+	}
+	return manifest, nil
+}
+
+// updateRelease updates a release in the cluster with an equivalent with the superseded APIs replaced or removed as
+// needed.
+// Logic extracted from https://github.com/helm/helm-mapkubeapis/blob/main/pkg/v3/release.go#L71-L94
+func updateRelease(originalRelease *release.Release, modifiedManifest string, config *action.Configuration) error {
+	originalRelease.Info.Status = release.StatusSuperseded
+	if err := config.Releases.Update(originalRelease); err != nil {
+		return errors.Wrapf(err, "[updateRelease] failed to update original release %v in namespace %v", originalRelease.Name, originalRelease.Namespace)
+	}
+
+	newRelease := originalRelease
+	newRelease.Manifest = modifiedManifest
+	newRelease.Info.Description = UpgradeDescription
+	newRelease.Info.LastDeployed = config.Now()
+	newRelease.Version = originalRelease.Version + 1
```

Comment:
Is this universally true? IIRC some of our charts rely on a specific versioning scheme to work - increasing the version by one for every chart could lead to cases where the versions modified here are seen as higher version than any future RCs, since we typically don't increment major versions very often.

Reply:
My understanding is that a release version _is not_ a chart version -- rather this is an internal Helm control to know what is the current version. It is also my understanding that this is always an integer.

Reply:
Ah - my bad, I misunderstood what that value was for. In that case, this seems ok, since future revisions should further bump this value by one (if I'm understanding correctly).